File: csharp/src/Contriwork.Notifications/Internal/QuietHours.cs
using System.Globalization;

namespace Contriwork.Notifications.Internal;

internal static class QuietHours
{
    /// <summary>Returns true when <paramref name="now"/> (or wall-clock now) falls inside the window.</summary>
    public static bool IsQuiet(QuietHoursConfig config, DateTimeOffset? now = null)
    {
        ArgumentNullException.ThrowIfNull(config);

        var tz = TimeZoneInfo.FindSystemTimeZoneById(config.Timezone);
        var moment = now ?? DateTimeOffset.UtcNow;
        var local = TimeZoneInfo.ConvertTime(moment, tz).TimeOfDay;
        var start = ParseHhMm(config.Start);
        var end = ParseHhMm(config.End);
        return start <= end
            ? local >= start && local < end
            : local >= start || local < end;
    }

    private static TimeSpan ParseHhMm(string value)
    {
        return TimeSpan.ParseExact(value, "h\\:mm", CultureInfo.InvariantCulture);
    }
}

File: typescript/src/internal/quiet-hours.ts
import type { QuietHoursConfig } from "../config";

function parseHhMm(value: string): number {
  const [h, m] = value.split(":");
  return Number(h ?? "0") * 60 + Number(m ?? "0");
}

/** Returns true when `now` (or wall-clock now) falls inside the window. */
export function isQuiet(config: QuietHoursConfig, now: Date = new Date()): boolean {
  const fmt = new Intl.DateTimeFormat("en-GB", {
    hour: "2-digit",
    minute: "2-digit",
    hour12: false,
    timeZone: config.timezone,
  });
  const parts = fmt.formatToParts(now);
  const hour = parts.find((p) => p.type === "hour")?.value ?? "00";
  const minute = parts.find((p) => p.type === "minute")?.value ?? "00";
  const nowMinutes = Number(hour) * 60 + Number(minute);

  const startMin = parseHhMm(config.start);
  const endMin = parseHhMm(config.end);

  if (startMin <= endMin) {
    return nowMinutes >= startMin && nowMinutes < endMin;
  }
  return nowMinutes >= startMin || nowMinutes < endMin;
}
